File: RecaptchaValidation/Models/RecaptchaRequestMessage.cs
using System.Runtime.Serialization;
using System.Web;

namespace RecaptchaValidation.Models;

[DataContract]
internal sealed class RecaptchaRequestMessage
{
    public string Path { get; }
    public string Secret { get; }
    public string Response { get; }
    public string RemoteIp { get; }

    public RecaptchaRequestMessage(string responseToken, string remoteIPAddress, string secretKey, string verifyUrl)
    {
        Path = verifyUrl;
        Secret = secretKey;
        Response = responseToken;
        RemoteIp = remoteIPAddress;
    }

    public static string GetVerificationUrl(RecaptchaRequestMessage requestMessage)
    {
        return $"{requestMessage.Path}?{HttpUtility.UrlPathEncode($"secret={requestMessage.Secret}&response={requestMessage.Response}&remoteip={requestMessage.RemoteIp}")}";
    }
    
}

File: RecaptchaValidation2.0/RecaptchaRequestMessage.fs
[<RequireQualifiedAccess>]
module internal RecaptchaRequestMessage

open System.Web

let fromOptions token remoteIp ({ VerifyUrl = url; Keys = keys }: RecaptchaOptions) =
    { Path = url
      Secret = keys.Secret
      Token = token
      RemoteIp = remoteIp }

let private encodeRequest request =
    let secret = sprintf "secret=%s" request.Secret
    let response = sprintf "response=%s" request.Token
    let remoteIp = sprintf "remoteip=%s" request.RemoteIp

    let result = sprintf "%s&%s&%s" secret response remoteIp
    HttpUtility.UrlPathEncode(result)

let verificationUrl request =
    let query = encodeRequest request
    sprintf "%s?%s" request.Path query